src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
-
-	if (nCode >= 0) {
-		bool suppress = false;
-		UINT msg      = wParam;
-		// Convert the hooked Windows mouse message into a DirectInput
-		// button index, and store the pressed state of the button.
-		bool down        = false;
-		unsigned int btn = 0;
-		switch (msg) {
-			case WM_LBUTTONDOWN:
-				down = true;
-			case WM_LBUTTONUP:
-				btn = 3;
-				break;
-			case WM_RBUTTONDOWN:
-				down = true;
-			case WM_RBUTTONUP:
-				btn = 4;
-				break;
-			case WM_MBUTTONDOWN:
-				down = true;
-			case WM_MBUTTONUP:
-				btn = 5;
-				break;
-			case WM_XBUTTONDOWN:
-				down = true;
-			case WM_XBUTTONUP:
-				btn = 5 + (mouse->mouseData >> 16);
-			default:
-				break;
-		}
-		if (btn > 0) {
-			suppress = handleMouseMessage(btn, down);
-			if (suppress) {
-				return 1;
-			}
-		}
-	}
-	return CallNextHookEx(gsw->hhMouse, nCode, wParam, lParam);
+	const QList< QVariant > ql{ QVariant::fromValue(DeviceTypes::MOUSE), button };
+	handleButton(ql, down) && button != 1;
 }
 
-BOOL CALLBACK GlobalShortcutWin::EnumDeviceObjectsCallback(LPCDIDEVICEOBJECTINSTANCE lpddoi, LPVOID pvRef) {
-	InputDevice *id             = static_cast< InputDevice * >(pvRef);
-	QString name                = QString::fromUtf16(reinterpret_cast< const ushort * >(lpddoi->tszName));
-	id->qhNames[lpddoi->dwType] = name;
-
-	if (Global::get().s.bDirectInputVerboseLogging) {
-		qWarning("GlobalShortcutWin: EnumObjects: device %s %s object 0x%.8lx %s",
-				 qPrintable(QUuid(id->guid).toString()), qPrintable(id->name),
-				 static_cast< unsigned long >(lpddoi->dwType), qPrintable(name));
+GlobalShortcutWin::Devices::iterator GlobalShortcutWin::addDevice(const HANDLE deviceHandle) {
+	RID_DEVICE_INFO deviceInfo;
+	UINT size = sizeof(deviceInfo);
+	if (GetRawInputDeviceInfo(deviceHandle, RIDI_DEVICEINFO, &deviceInfo, &size) <= 0) {
+		return m_devices.end();
 	}
 
-	return DIENUM_CONTINUE;
-}
+	Device device;
 
-BOOL GlobalShortcutWin::EnumDevicesCB(LPCDIDEVICEINSTANCE pdidi, LPVOID pContext) {
-	GlobalShortcutWin *cbgsw = static_cast< GlobalShortcutWin * >(pContext);
-	HRESULT hr;
+	std::stringstream nameStream;
+	nameStream << std::uppercase << std::hex << std::setfill('0');
+	nameStream << '[';
+	nameStream << std::setw(4) << deviceInfo.hid.dwVendorId;
+	nameStream << ':';
+	nameStream << std::setw(4) << deviceInfo.hid.dwProductId;
+	nameStream << ']';
 
-	QString name  = QString::fromUtf16(reinterpret_cast< const ushort * >(pdidi->tszProductName));
-	QString sname = QString::fromUtf16(reinterpret_cast< const ushort * >(pdidi->tszInstanceName));
+	device.prefix = nameStream.str() + ':';
 
-	InputDevice *id = new InputDevice;
-
-	id->pDID = nullptr;
+	if (GetRawInputDeviceInfo(deviceHandle, RIDI_DEVICENAME, nullptr, &size) == 0) {
+		std::wstring name;
+		name.resize(size);
+		// E.g. "\\?\HID#VID_045E&PID_02FF&IG_00#8&15c9c520&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}"
+		if (GetRawInputDeviceInfo(deviceHandle, RIDI_DEVICENAME, &name[0], &size) > 0) {
```

Comment:
For Strings the same applies as for vectors (at least I am pretty sure): You access the underlaying array via `data()`:
```suggestion
		if (GetRawInputDeviceInfo(deviceHandle, RIDI_DEVICENAME, name.data(), &size) > 0) {
```

And just to make it more clear that `size` is always equals the size of the passed string, I would pass `name.size()` instead of `size`.

Reply:
You can't, the function wants a parameter.

Reply:
As for `data()`, we cannot do that either because it returns a `const` pointer.

Reply:
> As for data(), we cannot do that either because it returns a const pointer.

Indeed. Apparently that is changed with cpp17 though (https://stackoverflow.com/questions/34155390/why-doesnt-stringdata-provide-a-mutable-char)

> You can't, the function wants a parameter.

Ah yes right :+1: